quickwit/quickwit-doc-mapper/src/default_doc_mapper/mapping_tree.rs

```diff
@@ -214,38 +184,61 @@ impl MappingLeaf {
     }
 }
 
-fn json_type_from_json_value(json_value: &JsonValue) -> JsonType {
-    match json_value {
-        JsonValue::Null => JsonType::Null,
-        JsonValue::Bool(_) => JsonType::Bool,
-        JsonValue::Array(_) => JsonType::Array,
-        JsonValue::Number(_) => JsonType::Number,
-        JsonValue::String(_) => JsonType::String,
-        JsonValue::Object(_) => JsonType::Object,
-    }
-}
-
 fn extract_json_val(
-    json_type: JsonType,
-    named_doc: &mut BTreeMap<String, Vec<JsonValue>>,
+    leaf_type: &LeafType,
+    named_doc: &mut BTreeMap<String, Vec<TantivyValue>>,
     field_path: &[&str],
     cardinality: Cardinality,
 ) -> Option<JsonValue> {
     let full_path = field_path.join(".");
     let vals = named_doc.remove(&full_path)?;
     let mut vals_with_correct_type_it = vals
         .into_iter()
-        .filter(|json_val| json_type_from_json_value(json_val) == json_type);
+        .filter_map(|value| value_to_json(value, leaf_type));
```

Comment:
```suggestion
        .flat_map(|value| value_to_json(value, leaf_type));
```

Reply:
`flat_map` is semantically better when flattening sequences:

```rust
let all_chars: Vec <_> = ["foo", "bar"].into_iter().flat_map(|word| word.chars()).collect();
```

`filter_map`

```rust
let even_octals: Vec <_> = [1, 2].into_iter().filter_map(|i| if i % 2 == 0 { Some(format!("{i:o}")) } else { None }).collect();
```